libdebug/cffi/ptrace_cffi_source.c

```diff
@@ -67,10 +78,317 @@ struct thread_status {
+        return 1;
+
+    return 0;
+}
+
+int get_remaining_hw_breakpoint_count(struct global_state *state, int tid)
+{
+    int i;
+    for (i = 0; i < 4; i++) {
+        unsigned long address = ptrace(PTRACE_PEEKUSER, tid, DR_BASE + i * DR_SIZE);
+
+        if (!address)
+            break;
+    }
+
+    return 4 - i;
+}
+
+int get_remaining_hw_watchpoint_count(struct global_state *state, int tid)
+{
+    return get_remaining_hw_breakpoint_count(state, tid);
+}
+#endif
+
+#ifdef ARCH_AARCH64
+struct user_hwdebug_state {
+    unsigned int dbg_info;
+	unsigned int pad;
+	struct {
+		unsigned long addr;
+		unsigned int ctrl;
+		unsigned int pad;
+	} dbg_regs[16];
+};
+
+int get_breakpoint_type(char type[2])
+{
+    if (type[0] == 'r') {
+        if (type[1] == 'w') {
+            return 3;
+        } else {
+            return 1;
+        }
+    } else if (type[0] == 'w') {
+        return 2;
+    } else if (type[0] == 'x') {
+        return 0;
+    } else {
+        return -1;
+    }
+}
+
+void install_hardware_breakpoint(struct hardware_breakpoint *bp)
+{
+    // find a free debug register
+    struct user_hwdebug_state state = {0};
+
+    struct iovec iov;
+    iov.iov_base = &state;
+    iov.iov_len = sizeof state;
+
+    unsigned long command = get_breakpoint_type(bp->type) == 0 ? NT_ARM_HW_BREAK : NT_ARM_HW_WATCH;
+
+    ptrace(PTRACE_GETREGSET, bp->tid, command, &iov);
+
+    int i;
+    for (i = 0; i < 16; i++) {
+        if (!state.dbg_regs[i].addr)
+            break;
+    }
+
+    if (i == 16) {
+        perror("No debug registers available");
+        return;
+    }
+
+    if (bp->type[0] == 'x') {
+        // Hardware breakpoint can only be of length 4
+        bp->len = 4;
+    }
+
+    unsigned int length = (1 << bp->len) - 1;
+    unsigned int condition = get_breakpoint_type(bp->type);
+    unsigned int control = (length << 5) | (condition << 3) | (2 << 1) | 1;
+
+    state.dbg_regs[i].addr = bp->addr;
+    state.dbg_regs[i].ctrl = control;
+
+    ptrace(PTRACE_SETREGSET, bp->tid, command, &iov);
+}
+
+void remove_hardware_breakpoint(struct hardware_breakpoint *bp)
+{
+    struct user_hwdebug_state state = {0};
+
+    struct iovec iov;
+    iov.iov_base = &state;
+    iov.iov_len = sizeof state;
+
+    unsigned long command = get_breakpoint_type(bp->type) == 0 ? NT_ARM_HW_BREAK : NT_ARM_HW_WATCH;
+
+    ptrace(PTRACE_GETREGSET, bp->tid, command, &iov);
+
+    int i;
+    for (i = 0; i < 16; i++) {
+        if (state.dbg_regs[i].addr == bp->addr)
+            break;
+    }
+
+    if (i == 16) {
+        perror("Breakpoint not found");
+        return;
+    }
+
+    state.dbg_regs[i].addr = 0;
+    state.dbg_regs[i].ctrl = 0;
+
+    ptrace(PTRACE_SETREGSET, bp->tid, command, &iov);
+}
+
+int is_breakpoint_hit(struct hardware_breakpoint *bp)
+{
+    siginfo_t si;
+
+    if (ptrace(PTRACE_GETSIGINFO, bp->tid, NULL, &si) == -1) {
+        return 0;
+    }
+
+    // Check that the signal is a SIGTRAP and the code is 0x4
```

Comment:
why this extra check w.r.t. amd64?

Reply:
Which extra check? This is completely different from how it's done on amd64